lib/cli/src/split_version.rs

```diff
@@ -0,0 +1,726 @@
+                    registry,
+                    &package_info.package,
+                    package_info.version.as_deref(),
+                    None,
+                    options.force_install,
+                );
+                if let Some(sp) = sp.take() {
+                    sp.clear();
+                }
+                let _ = std::io::stdout().flush();
+
+                match result {
+                    Ok((_, package_dir)) => Ok(RunWithPathBuf {
+                        path: package_dir,
+                        options,
+                    }),
+                    Err(e) => Err(anyhow::anyhow!(
+                        "could not install {}: {e}",
+                        package_info.package
+                    )),
+                }
+            }
+        }
+    }
+}
+
+/// Package specifier in the format of `package@version:command`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SplitVersionPackage {
+    /// namespace/package
+    pub package: String,
+    /// version: latest | 1.0.2
+    pub version: Option<ParsedPackageVersion>,
+    /// :command
+    pub command: Option<String>,
+}
+
+/// Command specifier in the format of `command@version`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SplitVersionCommand {
+    /// Command of the `command`
+    pub command: String,
+    /// version: latest | 1.0.2
+    pub version: Option<ParsedPackageVersion>,
+}
+
+/// What type of package should be run
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SplitVersionInner {
+    /// Run a file
+    File {
+        /// Path to the file
+        path: String,
+    },
+    /// Run a URL
+    Url(url::Url),
+    /// Run a package
+    Package(SplitVersionPackage),
+    /// Run a command
+    Command(SplitVersionCommand),
+}
+
+/// Package version specifier
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ParsedPackageVersion {
+    /// version 1.0.2
+    Version(semver::Version),
+    /// version latest
+    Latest,
+}
+
+impl fmt::Display for ParsedPackageVersion {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ParsedPackageVersion::Latest => write!(f, "latest"),
+            ParsedPackageVersion::Version(v) => write!(f, "{}", v),
+        }
+    }
+}
+
+impl SplitVersionInner {
+    /// Try parsing `s` as a URL
+    pub fn try_parse_url(s: &str) -> Result<Self, SplitVersionError> {
+        let url = url::Url::parse(s).map_err(|e| SplitVersionError::InvalidUrl(format!("{e}")))?;
+        if !(url.scheme() == "http" || url.scheme() == "https") {
+            return Err(SplitVersionError::InvalidUrl(format!(
+                "invalid scheme {}",
+                url.scheme()
+            )));
+        }
+        Ok(Self::Url(url))
+    }
+
+    /// Try parsing `s` as a package (`namespace/user`) or a file (`./namespace/user`)
+    pub fn try_parse_package(s: &str) -> Result<Self, SplitVersionError> {
+        let re1 = regex::Regex::new(r#"(.*)/(.*)@(.*):(.*)"#).unwrap();
```

Comment:
These regexes have lot's of problems.

First, they'll need to handle the whole input, like `"^ ..... $`.
Then, `(.*)/...`  the `.*` is greedy, so it would also include slashes if the later parts match.
You'll want either a non-greedy match `(.*?)` , or be more explicit, by excluding slashes, like `[^/]+`. 
Meaning: everything except a slash, and at least once.

But actually you'll want to be more specific, like `[a-zA-Z0-9\-_]+`.

This is why I said we really need a proper parser for this...

Reply:
The Regex just do the rough parsing. The input is checked for alphanumeric and slashes later on (a string `my/package/version` would pass the Regex but get discarded later on).

Reply:
Okay, this is now just one regex.